File: src/lib/utils/validation.ts
/**
 * Input validation utilities
 * Provides validation functions for user inputs
 */

/**
 * Validate search query
 */
export function validateSearchQuery(query: string): { valid: boolean; error?: string } {
  if (!query || typeof query !== "string") {
    return { valid: false, error: "Search query is required" };
  }

  const trimmed = query.trim();
  if (trimmed.length === 0) {
    return { valid: false, error: "Search query cannot be empty" };
  }

  if (trimmed.length > 200) {
    return { valid: false, error: "Search query is too long (max 200 characters)" };
  }

  // Check for potentially malicious patterns
  if (/[<>{}[\]\\]/.test(trimmed)) {
    return { valid: false, error: "Search query contains invalid characters" };
  }

  return { valid: true };
}

/**
 * Sanitize string input
 */
export function sanitizeString(input: string): string {
  return input
    .trim()
    .replace(/[<>{}[\]\\]/g, "") // Remove potentially dangerous characters
    .slice(0, 1000); // Limit length
}

/**
 * Validate numeric ID
 */
export function validateId(id: unknown): { valid: boolean; value?: number; error?: string } {
  if (typeof id !== "number" && typeof id !== "string") {
    return { valid: false, error: "ID must be a number" };
  }

  const numId = typeof id === "string" ? parseInt(id, 10) : id;

  if (isNaN(numId) || numId <= 0 || !Number.isInteger(numId)) {
    return { valid: false, error: "ID must be a positive integer" };
  }

  return { valid: true, value: numId };
}

/**
 * Validate date string
 */
export function validateDate(dateString: string): { valid: boolean; error?: string } {
  if (!dateString || typeof dateString !== "string") {
    return { valid: false, error: "Date is required" };
  }

  const date = new Date(dateString);
  if (isNaN(date.getTime())) {
    return { valid: false, error: "Invalid date format" };
  }

  return { valid: true };
}

/**
 * Validate URL
 */
export function validateUrl(url: string): { valid: boolean; error?: string } {
  if (!url || typeof url !== "string") {
    return { valid: false, error: "URL is required" };
  }

  try {
    new URL(url);
    return { valid: true };
  } catch {
    return { valid: false, error: "Invalid URL format" };
  }
}

File: src-tauri/src/commands/validation.rs
//! Input validation utilities for Rust commands
//! Mirrors the validation.ts patterns from the frontend

/// Validate that an ID is a non-zero integer (positive for API entries, negative for manual entries)
pub fn validate_id(id: i64) -> Result<(), String> {
    if id == 0 {
        return Err("ID cannot be zero".to_string());
    }
    Ok(())
}

/// Validate search query string
#[allow(dead_code)]
pub fn validate_search_query(query: &str) -> Result<(), String> {
    if query.trim().is_empty() {
        return Err("Search query cannot be empty".to_string());
    }

    if query.len() > 200 {
        return Err("Search query is too long (max 200 characters)".to_string());
    }

    // Check for potentially malicious patterns
    if query.contains('<')
        || query.contains('>')
        || query.contains('{')
        || query.contains('}')
        || query.contains('[')
        || query.contains(']')
        || query.contains('\\')
    {
        return Err("Search query contains invalid characters".to_string());
    }

    Ok(())
}

/// Validate date string format (YYYY-MM-DD)
pub fn validate_date(date: &str) -> Result<(), String> {
    use chrono::Datelike;
    let parsed = chrono::NaiveDate::parse_from_str(date, "%Y-%m-%d")
        .map_err(|_| "Enter a valid date in YYYY-MM-DD format".to_string())?;
    if parsed.format("%Y-%m-%d").to_string() != date || !(1900..=2100).contains(&parsed.year()) {
        return Err("Enter a date in YYYY-MM-DD format between 1900 and 2100".into());
    }
    Ok(())
}

/// Validate URL (for opening external links)
#[allow(dead_code)]
pub fn validate_url(url: &str) -> Result<(), String> {
    if url.is_empty() {
        return Err("URL is required".to_string());
    }

    // Only allow https:// URLs for security
    if !url.starts_with("https://") {
        return Err("Only HTTPS URLs are allowed".to_string());
    }

    // Basic URL format validation (check for valid structure)
    if url.len() < 8 || !url.contains('.') {
        return Err("Invalid URL format".to_string());
    }

    // Check for dangerous patterns
    if url.contains("javascript:") || url.contains("data:") || url.contains("<script") {
        return Err("URL contains potentially dangerous content".to_string());
    }

    Ok(())
}

/// Validate port number
pub fn validate_port(port: u16) -> Result<(), String> {
    if port == 0 {
        return Err("Port cannot be zero".to_string());
    }
    Ok(())
}
